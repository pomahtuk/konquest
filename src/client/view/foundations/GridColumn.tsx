import React, { ReactElement, useContext, HTMLAttributes, ReactNode } from "react";
import { styled } from "styletron-react";
import { ThemeContext } from "../themes/ThemeProvider";

export type GridSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | "half" | "full";

export type GridColumnProps = {
  children: ReactNode;
  size?: GridSize;
  sizeMedium?: GridSize;
  sizeLarge?: GridSize;
  sizeHuge?: GridSize;
  offset?: GridSize;
  offsetMedium?: GridSize;
  offsetLarge?: GridSize;
  offsetHuge?: GridSize;
  className?: string;
  attributes?: HTMLAttributes<HTMLDivElement>;
  bleed?: boolean;
};

const toRealValue = (val?: GridSize): GridSize | undefined => (val === "half" ? 6 : val === "full" ? 12 : val);

const GridColumn = ({
  size,
  sizeMedium,
  sizeLarge,
  sizeHuge,
  children,
  className,
  attributes,
  offset,
  offsetMedium,
  offsetLarge,
  offsetHuge,
  bleed
}: GridColumnProps): ReactElement => {
  const theme = useContext(ThemeContext);
  // prepare variables
  const realSize = toRealValue(size) || 12;
  const realOffset = toRealValue(offset);
  const realSizeMedium = toRealValue(sizeMedium);
  const realSizeLarge = toRealValue(sizeLarge);
  const realSizeHuge = toRealValue(sizeHuge);
  const realOffsetMedium = toRealValue(offsetMedium);
  const realOffsetLarge = toRealValue(offsetLarge);
  const realOffsetHuge = toRealValue(offsetHuge);
  // prepare styled component
  const GridColumnStyled = styled("div", {
    boxSizing: "border-box",
    padding: bleed ? "0" : `${theme.units.medium} calc(${theme.units.medium}/2) 0 calc(${theme.units.medium}/2)`,
    flexBasis: `calc(100% / (12/${realSize}))`,
    width: `calc(100% / (12/${realSize}))`,
    maxWidth: `calc(100% / (12/${realSize}))`,
    ...(realOffset ? { marginLeft: `calc(100% / (12/${realOffset}))` } : {}),
    [theme.media.medium]: {
      padding: bleed ? "0" : `${theme.units.large} calc(${theme.units.large}/2) 0 calc(${theme.units.large}/2)`,
      ...(realSizeMedium ? { width: `calc(100% / (12/${realSizeMedium}))` } : {}),
      ...(realOffsetMedium ? { marginLeft: `calc(100% / (12/${realOffsetMedium}))` } : {})
    },
    [theme.media.large]: {
      padding: bleed ? "0" : `${theme.units.larger} calc(${theme.units.larger}/2) 0 calc(${theme.units.larger}/2)`,
      ...(realSizeLarge ? { width: `calc(100% / (12/${realSizeLarge}))` } : {}),
      ...(realOffsetLarge ? { marginLeft: `calc(100% / (12/${realOffsetLarge}))` } : {})
    },
    [theme.media.huge]: {
      padding: bleed ? "0" : `${theme.units.largest} calc(${theme.units.largest}/2) 0 calc(${theme.units.largest}/2)`,
      ...(realSizeHuge ? { width: `calc(100% / (12/${realSizeHuge}))` } : {}),
      ...(realOffsetHuge ? { marginLeft: `calc(100% / (12/${realOffsetHuge}))` } : {})
    }
  });
  return (
    <GridColumnStyled {...attributes} className={className}>
      {children}
    </GridColumnStyled>
  );
};

export default GridColumn;
