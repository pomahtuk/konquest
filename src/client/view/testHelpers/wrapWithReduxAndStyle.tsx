import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
