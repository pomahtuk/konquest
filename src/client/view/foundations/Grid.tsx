import React, { ReactElement, useContext, HTMLAttributes, ReactNode } from "react";
import styled from "@emotion/styled";
import { ThemeContext } from "../themes/ThemeProvider";
import GridColumn from "./GridColumn";

export interface GridProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode;
  align?: "center" | "end";
  justify?: "space-between" | "center" | "space-around";
  bleed?: boolean;
  reversed?: boolean;
  className?: string;
}

export type GridSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | "half" | "full";

const Grid = ({ children, className, align, justify, bleed, reversed, ...rest }: GridProps): ReactElement => {
  const theme = useContext(ThemeContext);
  const GridStyled = styled.div({
    boxSizing: "border-box",
    clear: "both" /* Fix for floated elements before the grid, that break due to negative margin */,
    display: "flex",
    flexFlow: "row wrap",
    justifyContent: "flex-start",
    ...(align ? { alignItems: align !== "center" ? `flex-${align}` : align } : {}),
    ...(justify ? { justifyContent: justify } : {}),
    ...(reversed ? { flexDirection: "row-reverse" } : {}),
    [theme.media.small]: {
      margin: bleed ? "0" : `calc(-1 * ${theme.units.medium}) calc(-1 * ${theme.units.medium}/2) 0 calc(-1 * ${theme.units.medium}/2)`
    },
    [theme.media.medium]: {
      margin: bleed ? "0" : `calc(-1 * ${theme.units.large}) calc(-1 * ${theme.units.large}/2) 0 calc(-1 * v${theme.units.large}/2)`
    },
    [theme.media.large]: {
      margin: bleed ? "0" : `calc(-1 * ${theme.units.larger}) calc(-1 * ${theme.units.larger}) 0 calc(-1 * ${theme.units.larger}/2)`
    },
    [theme.media.huge]: {
      margin: bleed ? "0" : `calc(-1 * ${theme.units.largest}) calc(-1 * ${theme.units.largest}) 0 calc(-1 * ${theme.units.largest}/2)`
    }
  });
  return (
    <GridStyled {...rest} className={className}>
      {bleed
        ? React.Children.map(children, (child: ReactNode) => {
            if (child && typeof child !== "string" && typeof child !== "number" && child !== true && (child as ReactElement).type === GridColumn) {
              return React.cloneElement(child as ReactElement, { bleed: true });
            }
            return child;
          })
        : children}
    </GridStyled>
  );
};

export default Grid;
