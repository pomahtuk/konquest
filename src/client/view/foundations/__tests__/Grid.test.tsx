import React from "react";
import { render } from "@testing-library/react";
import { matchers } from "jest-emotion";

expect.extend(matchers);

import wrapWithReduxAndStyle from "../../testHelpers/wrapWithReduxAndStyle";
import Grid from "../Grid";
import GridColumn from "../GridColumn";

const gridDemoText = "grid";

describe("<Grid />", (): void => {
  it("renders <Grid /> component", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <Grid>
          <GridColumn>1</GridColumn>
        </Grid>
      )
    );
    expect(container).toMatchSnapshot();
  });

  it("renders <Grid /> with default styles", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid>{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).not.toHaveStyleRule("align-items", "center");
    expect(element).not.toHaveStyleRule("align-items", "flex-end");
    expect(element).not.toHaveStyleRule("justify-content", "space-between");
    expect(element).not.toHaveStyleRule("justify-content", "center");
    expect(element).not.toHaveStyleRule("justify-content", "space-around");
  });

  it("renders <Grid /> with align center", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid align="center">{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).toHaveStyleRule("align-items", "center");
  });

  it("renders <Grid /> with align end", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid align="end">{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).toHaveStyleRule("align-items", "flex-end");
  });

  it("renders <Grid /> with justify space-between", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid justify="space-between">{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).toHaveStyleRule("justify-content", "space-between");
  });

  it("renders <Grid /> with justify center", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid justify="center">{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).toHaveStyleRule("justify-content", "center");
  });

  it("renders <Grid /> with justify space-around", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid justify="space-around">{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).toHaveStyleRule("justify-content", "space-around");
  });

  it("renders <Grid /> with reversed order", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid reversed>{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element).toHaveStyleRule("flex-direction", "row-reverse");
  });

  it("renders <Grid /> in bleed variant", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <Grid bleed>
          <GridColumn>{gridDemoText}</GridColumn>
          {222}
        </Grid>
      )
    );

    expect(container).toMatchSnapshot();
  });

  it("renders <Grid /> with additional className", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Grid className="some">{gridDemoText}</Grid>));
    const element = getByText(gridDemoText);
    expect(element.classList.contains("some")).toBe(true);
  });
});
