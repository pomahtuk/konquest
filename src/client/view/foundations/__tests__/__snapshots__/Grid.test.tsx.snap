// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> renders <Grid /> component 1`] = `
.emotion-1 {
  box-sizing: border-box;
  clear: both;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

@media screen and (max-width:575px) {
  .emotion-1 {
    margin: calc(-1 * 8px) calc(-1 * 8px/2) 0 calc(-1 * 8px/2);
  }
}

@media screen and (min-width:576px) {
  .emotion-1 {
    margin: calc(-1 * 16px) calc(-1 * 16px/2) 0 calc(-1 * v16px/2);
  }
}

@media screen and (min-width:992px) {
  .emotion-1 {
    margin: calc(-1 * 24px) calc(-1 * 24px) 0 calc(-1 * 24px/2);
  }
}

@media screen and (min-width:1200px) {
  .emotion-1 {
    margin: calc(-1 * 32px) calc(-1 * 32px) 0 calc(-1 * 32px/2);
  }
}

.emotion-0 {
  box-sizing: border-box;
  padding: 8px calc(8px/2) 0 calc(8px/2);
  -webkit-flex-basis: calc(100% / (12/12));
  -ms-flex-preferred-size: calc(100% / (12/12));
  flex-basis: calc(100% / (12/12));
  width: calc(100% / (12/12));
  max-width: calc(100% / (12/12));
}

@media screen and (min-width:576px) {
  .emotion-0 {
    padding: 16px calc(16px/2) 0 calc(16px/2);
  }
}

@media screen and (min-width:992px) {
  .emotion-0 {
    padding: 24px calc(24px/2) 0 calc(24px/2);
  }
}

@media screen and (min-width:1200px) {
  .emotion-0 {
    padding: 32px calc(32px/2) 0 calc(32px/2);
  }
}

<div>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-0"
    >
      1
    </div>
  </div>
</div>
`;

exports[`<Grid /> renders <Grid /> in bleed variant 1`] = `
.emotion-1 {
  box-sizing: border-box;
  clear: both;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

@media screen and (max-width:575px) {
  .emotion-1 {
    margin: 0;
  }
}

@media screen and (min-width:576px) {
  .emotion-1 {
    margin: 0;
  }
}

@media screen and (min-width:992px) {
  .emotion-1 {
    margin: 0;
  }
}

@media screen and (min-width:1200px) {
  .emotion-1 {
    margin: 0;
  }
}

.emotion-0 {
  box-sizing: border-box;
  padding: 0;
  -webkit-flex-basis: calc(100% / (12/12));
  -ms-flex-preferred-size: calc(100% / (12/12));
  flex-basis: calc(100% / (12/12));
  width: calc(100% / (12/12));
  max-width: calc(100% / (12/12));
}

@media screen and (min-width:576px) {
  .emotion-0 {
    padding: 0;
  }
}

@media screen and (min-width:992px) {
  .emotion-0 {
    padding: 0;
  }
}

@media screen and (min-width:1200px) {
  .emotion-0 {
    padding: 0;
  }
}

<div>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-0"
    >
      grid
    </div>
    222
  </div>
</div>
`;
